File: typingspeed.rb
require 'terminfo'
require 'colorize'

TIME_LIMIT = 1 # Seconds
CHARS_PER_WORD = 5
height, width = TermInfo.screen_size

puts "##### Typing Speed Test! #####".blue
puts "##### By: Nithin Singhal #####".blue
puts "### Press 'Enter' to start ###".blue
gets

words_list = Array.new
File.readlines('commonwords.txt', chomp: true).each do |word|
  words_list << word
end

mispelled_words = Array.new
correct_words_typed = 0
total_words_printed = 0
total_words_typed = 0
total_chars_typed = 0

t_start = Time.now
while (Time.now - t_start).to_i < TIME_LIMIT do
  line_length = 0
  line = ""
  print("  ")
  while true do
    word = words_list[rand(1000)] + " "
    # Prevent duplicate words from appearing in line
    while line.include? word do
      word = words_list[rand(1000)] + " "
    end

    # Stop adding words when line becomes too long
    if line_length + word.length >= width - 1
      break
    end

    line += word
    line_length += word.length
  end
  print (line + "\r").cyan
  print "> "
  input_line = gets.chomp
  puts

  total_chars_typed += input_line.length

  line = line.split(" ")
  input_line = input_line.split(" ")
  total_words_printed += line.length
  total_words_typed += input_line.length
  line.each do |word|
    if input_line.include? word
      correct_words_typed += 1
    else
      mispelled_words << word
    end
    #puts "Correct words typed: " + correct_words_typed.to_s
    #puts "Total words printed: " + total_words_printed.to_s
  end
end
t_end = Time.now
time_taken = (t_end - t_start).to_f
puts "Time taken: #{format("%.2f", time_taken)}s".green
print("Mispelled words:".red)
if mispelled_words.empty?
  puts " None!".green
else
  mispelled_words.each do |word|
    print (" " + word).red
  end
  puts
end

puts "Accuracy: #{format("%1.2f", (correct_words_typed.to_f / total_words_printed * 100))}%".blue
puts "Words Per Minute (WPM): #{format("%.2f", (total_chars_typed.to_f / CHARS_PER_WORD / (time_taken / 60)))}".blue
puts "Characters Per Second (CPS): #{format("%.2f", (total_chars_typed.to_f / time_taken))}".blue

File: typingspeed.py
import os
import random
import time
from termcolor import colored

TIME_LIMIT = 60 # Seconds
CHARS_PER_WORD = 5
width, height = os.get_terminal_size()
width = min(width, 80)

input(colored("""##### Typing Speed Test! #####\n"""
              """##### By: Nithin Singhal #####\n"""
              """### Press 'Enter' to start ###\n""", "blue"))

with open("commonwords.txt", "r") as f:
    words_list = f.read().split()
f.close()

def randWord():
    return words_list[random.randint(0, len(words_list) - 1)]

mispelled_words = []
correct_words_typed = 0
total_words_printed = 0
total_words_typed = 0
total_chars_typed = 0

t_start = time.time()
while time.time() < t_start + TIME_LIMIT:
    line_length = 0
    line = ""
    print("  ", end="")
    while True:
        word = randWord() + " "
        # Prevent duplicate words from appearing in line
        while word in line:
            word = randWord() + " "

        # Stop adding words when line becomes too long
        if line_length + len(word) >= width - 1:
            break

        line += word
        line_length += len(word)
    print(colored(line, "cyan"), end="\r")
    input_line = input("> ")
    print()

    total_chars_typed += len(input_line)

    line = line.split()
    input_line = input_line.split()
    total_words_printed += len(line)
    total_words_typed += len(input_line)

    for word in line:
        if word in input_line:
            correct_words_typed += 1
        else:
            mispelled_words.append(word)

t_end = time.time()
time_taken = t_end - t_start
print(colored("Time taken: " + "{:.2f}s".format(time_taken), "green"))
print(colored("Mispelled words:", "red"), end="")
if mispelled_words:
    for word in mispelled_words:
        print(colored(" " + word, "red"), end="")
    print()
else:
    print(colored(" None!", "green"))

print(colored("Accuracy: " + "{0:.2f}%".format(correct_words_typed / total_words_printed * 100), "blue"))
print(colored("Words Per Minute (WPM): " + "{:.2f}".format(total_chars_typed / CHARS_PER_WORD / (time_taken / 60)), "blue"))
print(colored("Characters Per Second (CPS): " + "{:.2f}".format(total_chars_typed / time_taken), "blue"))
